include/common/patterns.h

```diff
@@ -81,9 +83,35 @@ constexpr bool is_no_pattern =
   !is_reduce<T> &&
   !is_iteration<T>;
 
+
+template <typename T>
+constexpr bool is_pattern = !is_no_pattern<T>;
+
 template <typename T>
 using requires_no_pattern = std::enable_if_t<is_no_pattern<T>,int>;
 
+template <typename T>
+using requires_pattern = std::enable_if_t<is_pattern<T>, int>;
+
+
+template <typename return_type, typename ... Ts>
+struct get_return{
```

Comment:
Probably `get_return` is not a good name for this meta-function.

Reply:
Renamed to get_return_type